src/Microsoft.ML.Transforms/BootstrapSampleTransform.cs

```diff
@@ -76,6 +76,12 @@ public BootstrapSampleTransform(IHostEnvironment env, Arguments args, IDataView
             _poolSize = args.PoolSize;
         }
 
+        public BootstrapSampleTransform(IHostEnvironment env, IDataView input, bool complement = false, uint? seed = null, bool shuffleInput = true, int poolSize = 1000)
```

Comment:
>bool complement = false, uint? seed = null, bool shuffleInput = true, int poolSize = 1000 [](start = 79, length = 89)

out of curiosity, why argument object is bad solution? why we want to flatten that object in set of parameters?
I'm also concern what we will have to support this defaults in two places, and it really error prone.

Reply:
so that transform can be neatly instantiated on one line. Otherwise for every transform a relevant "Argument" object is needed to be created first which is also another option.

---
In reply to: [197909492](https://github.com/dotnet/machinelearning/pull/405#discussion_r197909492) [](ancestors = 197909492)

Reply:
I can see why it can be helpful in case if you have only Column object inside argument class, you add one more constructor which is nice to use. But in this particular case, i don't see any advantages of using this constructor, and it makes code maintainability much harder.
Also compare this constructor with  Cathash constructor.
In cathash you hide all other parameters from argument class other than columns, and here you expose them.
Can we be consistent? 

I don't mind to have constructor which will take only Columns information and will use default parameters for internal argument class, but make sure to proper comment which indicates what if user needs more, it need to use Arguments class

---
In reply to: [197912065](https://github.com/dotnet/machinelearning/pull/405#discussion_r197912065) [](ancestors = 197912065,197909492)

Reply:
but again, feel free to bring heavy artillery (@TomFinley)

---
In reply to: [197921908](https://github.com/dotnet/machinelearning/pull/405#discussion_r197921908) [](ancestors = 197921908,197912065,197909492)

Reply:
For this case, definitely constructor can be removed. But I have a concern regarding if user has to instantiate an `Argument` object for a very simple case e.g. for all default values.

Another option could be
``` C#
public BootstrapSampleTransform(IHostEnvironment env, IDataView input)
: this(env, new Arguments() { }, input)
{
}
```

Yes, for `CatHash` more parameters can be added. Waiting for more comments on this.

Reply:
TBH I actually rather liked the constructor as it stands -- I do not consider the proposed replacement `BootstrapSampleTransform(IHostEnvironment env, IDataView input)` acceptable because the bootstrap sampler I expect is one of those things that, when it is used, will be used in a pair: first to get the in-bag sample (for training), then to get the out of bag sample (for validation and testing). Whether we expose the other things as parameters or not, I am less certain about the utility of these.

@Ivanidzo4ka , is the concern with having a convenience constructor with default arguments, that the defaults could diverge? If that is the case, would perhaps prefer constants (either directly as `private const` in the transform class itself, or perhaps more ideally as `public const` in a nested `private static` class named `Defaults` or something).

If the point is rather that there should be exactly one way to do this, I'm not entirely certain I agree -- the arguments object may need to exist for the command line parser and entry-points which (being reflection driven) require total uniformity in how things are instantiated, but this bears very little resemblance to how someone, if trying to write a .NET API, would actually write code. If it is a small "shim" as it is now over the arguments, I'm not sure I see the harm, and I see a considerable benefit.

Reply:
I just don't see much reason to have constructor which duplicate other constructor with one difference: instead of one object we accept all the fields of this object.
I just don't understand why it's better and what convenience it brings. 
I'm perfectly fine with something like

class(main_option)
class(main_option, secondary_parameter)
i just don't understand why you want to have 
class(A,B,C,D,E,F)
and 
class(argument_object) where argument_object has fields A,B,C,D,E,F

---
In reply to: [197983331](https://github.com/dotnet/machinelearning/pull/405#discussion_r197983331) [](ancestors = 197983331)

Reply:
I disagree. Where they can be applied, the new idioms are clearly easier to use and understand.

---
In reply to: [198273557](https://github.com/dotnet/machinelearning/pull/405#discussion_r198273557) [](ancestors = 198273557,197983331)

Reply:
This is based on the convenience you get in calling constructor. In my point of view, calling

``` C#
new BootstrapSampleTransform(env, input, Complement,  Seed, ShuffleInput, PoolSize);
```

is more convenient and elegant than calling 

``` C#
new BootstrapSampleTransform(env, input, new Arguments() { Complement = complement, Seed = seed, ShuffleInput = shuffleInput, PoolSize = poolSize });
```


---
In reply to: [198273557](https://github.com/dotnet/machinelearning/pull/405#discussion_r198273557) [](ancestors = 198273557,197983331)